wpep0003.md

```diff
@@ -0,0 +1,173 @@
+| Key | Value |
+| ---: | :--- |
+| **Title** | Improve acl system scalibility |
+| **Number** | 0003 |
+| **Date** | 2019-12-10 |
+| **Authors** | Mehdi Abaakouk <mabaakouk@wazo.io> |
+| **Status** | proposed |
+
+## Abstract
+
+Redesign the way services checks token, alcs and tenant ownership
+
+## Motivation
+
+Currently Wazo-auth is heavily solicitated just for checking token validity, acls and tenant ownership relative to the supplied token.
+
+### First issue, A service do three API calls to wazo-auth before it starts to handle a request.
+
+The user get a call::
+
+    GET /calls/{call_id} HTTP/1.1
+    X-Auth-Token: {token}
+
+Then wazo-calld check the token on wazo-auth and acl with::
+
+    HEAD /token/{token}?scope=calld.calls.create
+
+Wazo-auth compares the scope with the user acls.
+
+Wazo-auth returns to wazo-calld::
+
+    204 No content
+
+Then Wazo-calld retrieves the tenant information on wazo-auth::
+
+    GET /token/{token}
+
+Wazo-auth returns the information::
+
+    200 OK
+    {
+        "data": {
+            "acls": [],
+            "auth_id": "string",
+            "expires_at": "string",
+            "issued_at": "string",
+            "metadata": {
+                "tenant_uuid": "string"
+            },
+            "session_uuid": "string",
+            "token": "string",
+            "utc_expires_at": "string",
+            "utc_issued_at": "string",
+        }
+    }
+
+Then Wazo-calld retrieves the tenant tree on wazo-auth to check if the user is authorized to see the call::
+
+    GET /tenants
+
+Wazo-auth returns the information::
+
+    200 Ok
+    {
+        "filtered": 0,
+        "items": [
+            {
+                "name": "string",
+                "uuid": "string"
+            }
+        ],
+        "total": 0
+    }
+
+
+Wazo-calld now checks if the user `tenant_uuid` is part of the return list to grant access to the call.
+
+Wazo-calld can now returns the call information to the user::
+
+    200 OK
+    {
+        ...
+    }
+
+
+### Second issue, token creation and acl with variables
+
+When we create a token, wazo-auth retrieves the policy for the user.
+
+If the policy contains a variable (eg: {voicemail_id}), wazo-auth will do
+additional and authenticated requests to wazo-confd (that will do the post check seen previously)
+
+Due to the previous issue, this means token creation takes 4 wazo-auth API calls
+
+It's not clear what the use-case behind this variable replacement. I didn't
+find it used in our engine or portal policies.
```

Comment:
Some additional note: this link creates deadlock between wazo-auth and wazo-confd. When the maximal number of connections of wazo-auth if reached, when Wazo-auth want to retrieve data from confd, confd is unable to check the service token validity because of the http connection pools exhaution and then timeout, making the acl check timeout too.